File: tests/unit/helpers/rpc-server-child.py
import json
import sys

scenario = sys.argv[1]


def emit(value: bytes) -> None:
    sys.stdout.buffer.write(value)
    sys.stdout.buffer.flush()


if scenario == "missing-initialization":
    sys.stdin.readline()
    raise SystemExit(0)

initialize = json.loads(sys.stdin.readline())
assert initialize["id"] == 0 and initialize["method"] == "initialize"
emit(b'{"id":0,"result":{}}\n')
initialized = json.loads(sys.stdin.readline())
assert initialized == {"method": "initialized"}
request = json.loads(sys.stdin.readline())
assert request["id"] == 1
assert request["method"] in ("hooks/list", "skills/list")
assert len(request["params"]["cwds"]) == 1
if request["method"] == "skills/list":
    assert request["params"]["forceReload"] is True

if scenario == "malformed-json":
    emit(b"{\n")
elif scenario == "invalid-utf8":
    emit(b"\xff\n")
elif scenario == "nan":
    emit(b'{"id":1,"result":NaN}\n')
elif scenario == "non-object":
    emit(b"[]\n")
elif scenario == "rpc-error":
    emit(b'{"id":1,"error":{"code":1}}\n')
elif scenario == "no-result":
    emit(b'{"id":1}\n')
elif scenario == "wrong-ids":
    emit(b'{"method":"notice"}\n{"id":true,"result":{"data":["wrong"]}}\n{"id":2,"result":{"data":["wrong"]}}\n{"id":1,"result":{"data":[]}}\n')
elif scenario == "eof":
    raise SystemExit(0)
elif scenario == "timeout":
    sys.stdin.read()
elif scenario in ("hooks-success", "skills-success"):
    emit(b'{"id":1,"result":{"data":[]}}\n')
else:
    raise SystemExit(f"unknown scenario: {scenario}")

sys.stdin.read()

File: tests/unit/helpers/rpc-server-child.ts
import assert from "node:assert/strict";
import { closeSync } from "node:fs";
import { createInterface } from "node:readline";

const scenario = process.argv[2];
assert.ok(scenario, "expected scenario argument");

const input = createInterface({ input: process.stdin, crlfDelay: Infinity });
const requests = input[Symbol.asyncIterator]();

function asObject(value: unknown): Record<string, unknown> {
  assert.equal(typeof value, "object");
  assert.notEqual(value, null);
  assert.ok(!Array.isArray(value));
  return value as Record<string, unknown>;
}

async function request(): Promise<Record<string, unknown>> {
  const next = await requests.next();
  assert.equal(next.done, false, "unexpected input EOF");
  return asObject(JSON.parse(next.value));
}

function emit(value: string | Uint8Array): void {
  process.stdout.write(value);
}

async function assertNoExtraRequestBytes(): Promise<void> {
  for await (const _line of requests) {
    assert.fail("unexpected request bytes after initialization response");
  }
}

async function main(): Promise<void> {
  const initialize = await request();
  assert.equal(initialize.id, 0);
  assert.equal(initialize.method, "initialize");

  if (scenario === "missing-initialization") {
    closeSync(process.stdout.fd);
    await assertNoExtraRequestBytes();
    return;
  }

  emit('{"id":0,"result":{}}\n');
  assert.deepEqual(await request(), { method: "initialized" });
  const list = await request();
  assert.equal(list.id, 1);
  assert.ok(list.method === "hooks/list" || list.method === "skills/list");
  const params = asObject(list.params);
  assert.ok(Array.isArray(params.cwds));
  assert.equal(params.cwds.length, 1);
  if (list.method === "skills/list") assert.equal(params.forceReload, true);

  if (scenario === "malformed-json") emit("{\n");
  else if (scenario === "invalid-utf8") emit(new Uint8Array([0xff, 0x0a]));
  else if (scenario === "nan") emit('{"id":1,"result":NaN}\n');
  else if (scenario === "non-object") emit("[]\n");
  else if (scenario === "rpc-error") emit('{"id":1,"error":{"code":1}}\n');
  else if (scenario === "no-result") emit('{"id":1}\n');
  else if (scenario === "wrong-ids") {
    emit(
      '{"method":"notice"}\n{"id":true,"result":{"data":["wrong"]}}\n{"id":2,"result":{"data":["wrong"]}}\n{"id":1,"result":{"data":[]}}\n',
    );
  } else if (scenario === "eof") {
    input.close();
    return;
  } else if (scenario === "timeout") {
    await assertNoExtraRequestBytes();
    return;
  } else if (scenario === "hooks-success" || scenario === "skills-success") {
    emit('{"id":1,"result":{"data":[]}}\n');
  } else {
    assert.fail(`unknown scenario: ${scenario}`);
  }

  await assertNoExtraRequestBytes();
}

await main();
